### ISSUE TITLE
Add a mechanism for self-declaring a DID's entity type

### ISSUE THREAD

**Issue by csuwildcat** [decentralized-identity/sidetree#521]
Given we will have DIDs for humans, organizations, machines, and other things, it would be helpful to have an optional mechanism by which DID creators can self-declare what type of thing a DID represents. One way to do this could be to add a type field to the create operation Anchor File section with a couple bytes of space for a value that mapped to entity types. That's just one idea on how to do this, but this would really help in being able to have a general grouping of different DID types in the system.

**Comment by OR13** [comment]
Can't be trusted... anyone can write dids of type "make you waste time"....there is no write side protection in sidetree.

You can build an index based on your own surveillance / visibility and bucket dids  / pre cache them... we do this for dids when we know we might loose connectivity... but this is 100% not a sidetree specific thing... 

For example:

Lets say I know did:exampe:123 is actually Azure AD Tenant 456, because I can see that... I build a table for Azure AD Tenant 456 that maps their AD Entity types to DIDs.... I don't publish that mapping on a ledger... because thats customer information... 

That tenant can ask for did lists by type:

POST /tenant/456/dids { type: organization } => customer specific data that nobody should be able to learn from crawling public resources like bitcoin or ipfs....

Then they might locally resolve all those dids before starting their run through an internet denied environment... When they encounter an entity with a DID, they can authenticate it from cached data, even without internet.... 

TL;DR; this would be a terrible feature to implement in sidetree, but it's a great feature to implement in an sdk, or other services... we do this already in our enterprise agent.

**Comment by OR13** [comment]
categories of "type" are metadata... 

"we kill people based on metadata." 
- Michael Hayden

we won't be able to scrub timing information from sidetree ever... thats metadata as well...

If you are really worried, you should pre-allocate a really large number of DIDs and keep them all alive, and try and disconnect things like updates from when you are required to make them...

For example, if you add healthcare.edv.example.com to your did document the same day you get onboarded to a new clinic... its now really easy to see when that clinic gets new patients...

metadata is poison, we should be very careful to minimize it.

**Comment by csuwildcat** [comment]
@OR13 let's talk - I can provide some info on what the issues/details are that make this a significant need at this level.

**Comment by csuwildcat** [comment]
Another way to do this, as Orie noted on a call, would be to reserve the first two characters of a DID suffix string for this type of category indicator that a DID controller could optionally use, but then it means that a classification may be applied without explicit action, which I don't like. Separating it could cost a couple extra bytes, but might be worth it.

**Comment by OR13** [comment]
@kdenhartog @troyronda  add your comments to this ticket.

**Comment by OR13** [comment]
@csuwildcat to add language to the spec about "implementers may add other properties that mutate what the didUniqueSuffix will be".

**Comment by kdenhartog** [comment]
My opinion on this is that I'm not opposed to it, but I don't think this level of classification belongs at this layer. We likely wouldn't want to support it's usage because of the metadata concerns described above, but I don't see blaring concerns with this. I do think we're going to want the didUniqueSuffix to be clearly identifiable though if it's going to support this. Also might want to check with @SmithSamuelM to make sure we aren't breaking self certifying identifier property in any way.

**Comment by csuwildcat** [comment]
It's definitely not breaking anything about self-certification, given any
additional values in the Suffix Data are implicitly part of the ID itself.
(aside: Sidetree was the first self-certifying type of ID before that
concept existed in the DID ecosystem, besides arguably did:key's simple
linkage, so I'm confident on this point)

**Comment by SmithSamuelM** [comment]
Not sure what you mean by suffix. After the did  method  there is the unique name string  and then there may any number of colon separated strings. But forcing an entity type semantic on one of those colon separated strings is not likely an acceptable mechanism. Also forcing an entity type in the unique name string is bad for privacy. Some optional did metadata Is likely better.

@csuwildcat. FYI. My original open reputation white paper in early 2015 advocated the use of self-certifying IDs  for decentralized identity.  I am pretty sure that pre dates Sidetree by some time ;) likewise the Evernym white paper I wrote in early 2016I advocates for self-certifying identifiers.  ;).  Indeed  my original intent in those white papers was that all  decentralized  identifiers should be self-certifying. Unfortunately the  community around the nascent DID standard  in late 2016 dropped tself-certifying as a requirement in order to be more open to any type of did method . Those advocating DIDs at that time may not have fully appreciated what Self-certifying meant. This decision to drop self-certifying as a requirement carried with it security risks that were also likely not appreciated at the time. I have spent a lot of effort lately re-evangelizing self-certifying Identifiers but in no way is it a recent idea in the did ecosystem. Indeed it was the incepting idea for the DID ecosystem albeit one that was largely ignored or misunderstood until recently.

**Comment by csuwildcat** [comment]
The proposal for the type string, which would not even happen in Sidetree itself, was a 4 byte value to be included in the Suffix Data object that makes up the root of trust values for the IDs, not some additional segment in the DID URI. Suffix data is the immutable, self-certifying component of the URI, thus whatever goes into it is inherently self-certifying.  All of this is detailed in the DID URI composition section of the spec: https://identity.foundation/sidetree/spec/#did-uri-composition

Sidetree was first developed in 2017, so I guess the concept of a 'self-certifying' ID existed before that, but was unknown to me, as I never heard about it or read anything about it while I was working on Sidetree. I made Sidetree 'self-certifying' without even knowing it because it just seemed obvious you would want the feature in an ID system (I am honestly stunned to hear this was not plainly obvious to anyone who worked on decentralized identity).

**Comment by csuwildcat** [comment]
I remember thinking in 2017: "I should probably just make this sucker the hash of the initial state, because then you can tie a lineage of deltas to it deterministically" <-- how is that not the point of all DID systems? Are there actually Methods that _don't_ do this? If so, I feel like we should we tell anyone who made a DID Method like that to burn it with fire.
